src/docbuilder/caching.rs

```diff
@@ -0,0 +1,112 @@
+use crate::utils::copy_dir_all;
+use anyhow::{Context as _, Result};
+use std::{
+    fs, io,
+    path::{Path, PathBuf},
+};
+use tracing::{debug, instrument, warn};
+
+/// move cache folder to target, falling back to copy + delete on error.
+fn move_or_copy<P: AsRef<Path> + std::fmt::Debug, Q: AsRef<Path> + std::fmt::Debug>(
+    source: P,
+    dest: Q,
+) -> io::Result<()> {
+    if let Some(parent) = dest.as_ref().parent() {
+        fs::create_dir_all(parent)?;
+    }
+    if let Err(err) = fs::rename(&source, &dest) {
+        warn!(
+            ?err,
+            ?source,
+            ?dest,
+            "could not move target directory, fall back to copy"
+        );
+        copy_dir_all(&source, &dest)?;
+        fs::remove_dir_all(&source)?;
+    }
+    Ok(())
+}
+
+/// artifact caching with cleanup
+#[derive(Debug)]
+pub(crate) struct ArtifactCache {
+    cache_dir: PathBuf,
+}
+
+impl ArtifactCache {
+    pub(crate) fn new(cache_dir: PathBuf) -> Result<Self> {
+        Ok(Self { cache_dir })
+    }
+
+    pub(crate) fn purge(&self) -> Result<()> {
+        fs::remove_dir_all(&self.cache_dir)?;
+        Ok(())
+    }
+
+    /// clean up a target directory.
+    ///
+    /// Should delete all things that shouldn't leak between
+    /// builds, so:
+    /// - doc-output
+    /// - ...?
+    #[instrument(skip(self))]
+    fn cleanup(&self, target_dir: &Path) -> Result<()> {
+        for item in fs::read_dir(target_dir)? {
+            // the first level of directories are the targets in most cases,
+            // delete their doc-directories
+            let item = item?;
+            let doc_dir = item.path().join("doc");
+            if doc_dir.is_dir() {
+                debug!(?doc_dir, "cache dir cleanup");
+                fs::remove_dir_all(doc_dir)?;
+            }
```

Comment:
This doesn't handle proc-macros correctly, their output goes under `target/doc`, not `target/x86_64-linux-gnu/doc`.

Reply:
omg our tests caught this!!! I'm so unused to have tests lmao
```
2023-03-10T11:44:05.395187Z DEBUG build_package{name="thiserror-impl" version="1.0.26" kind=CratesIo}: docs_rs::docbuilder::rustwide_builder: rename .workspace/builds/thiserror-impl-1.0.26/target/doc to .workspace/builds/thiserror-impl-1.0.26/target/x86_64-unknown-linux-gnu/doc
the test failed: File exists (os error 17)
  caused by: File exists (os error 17)
disabled backtrace
thread 'docbuilder::rustwide_builder::tests::test_proc_macro' panicked at 'the test failed', src/docbuilder/rustwide_builder.rs:1108:9
```

Reply:
I changed the code so it also deletes `doc` in the target-root.